src/components/questionnaire.js

```diff
@@ -0,0 +1,224 @@
+
+            if (!validationObj.test(input)) {
+                validationErrors.push({
+                    type: validationType,
+                    input: input,
+                    message: validationObj.message
+                })
+            }
+		}
+
+		var newErrors = this.state.validationErrors;
+		newErrors = newErrors.concat(validationErrors);
+
+		if (validationErrors.length) {
+			this.setState({
+				validationErrors: newErrors
+			})
+		}
+		else {
+			this.setState({
+				validationErrors: []
+			});
+		}	
+		return validationErrors;
+	};
+
+	handleButtonSelect = (questionId, answerValue, answerInfo) => {
+		let activeQuestion = this.getMessageById(questionId);
+
+		if (!answerInfo) {
+			answerInfo = {};
+
+			// The answer must already be in the state
+			const answerObjExists = Object.hasOwnProperty.call(this.state.answers, answerValue);
+			const answer = answerObjExists && !!this.state.answers[answerValue] 
+				? this.state.answers[answerValue] : null;
+
+			if (!answerObjExists || !answer) {
+				return;
+			}
+
+			// Set the step = to the value in the state
+			answerInfo.text = answer;
+
+			//What is the next Step
+			// A custom method or the options should figure out what this is
+			var isMethod = typeof activeQuestion.options === 'function';
+
+			if (isMethod) {
+				answerInfo.nextStep = activeQuestion.options(this.state.answers);
+			}
+		}
+
+		// Set the active question to answered
+		this.setLogQuestionToAnswered(questionId);
+
+		// Push the answer to the question log
+		this.addToLog({
+			text: answerInfo.text,
+			type: 'answer'
+		});
+
+		// Add the answer to the answer list
+		this.setAnswer(activeQuestion.key, answerValue);
+
+		// Go to the next question, if it exists
+		this.getNextMessage(answerInfo.nextStep);
+	};
+
+	handleFormSubmit = (e) => {
+		console.log(e);
+
+	};
+	
+	handleTextInputChange = (questionKey, validationTypes, changeEvent) => {
+		const answers = this.state.answers;
+		const zipStr = changeEvent.target.value;
+		const validationErrors = this.getValidationErrors(validationTypes, zipStr, questionKey);
+
+		if (validationErrors.length === 0) {
+			answers[questionKey] = zipStr;
+		}
+		else {
+			answers[questionKey] = null;
+		}
+
+		this.setState({
+			answers: answers
+		});
+	};
+
+    restartQuestionnaire = () => {
+        var initialState = this.getInitialState();
+        this.setState(initialState, () => {
+            this.startQuestionnaire();
+        });
+	};
+
+	setAnswer = (key, answer) => {
+		let answers = this.state.answers;
+		answers[key] = answer;
+
+		this.setState({
+			answers: answers
+		})
+	}
+
+	setLogQuestionToAnswered = (id) => {
+		const messageLog = this.state.log;
+		let logItemToUpdateIndex;
+		let logItemToUpdate;
+
+		for (let i = 0, len = messageLog.length; i < len; i++) {
+			const logItem = Object.assign({}, messageLog[i]);
+			const idExists = Object.hasOwnProperty.call(logItem, 'id');
+			const idMatches = logItem.id === id;
```

Comment:
So, you're checking if the `logItem.id` exists, and then on the next line you're using it. I think you can kill the existence check, since if the property is undefined, you'd be comparing undefined to the supplied `id` parameter, and it'd be false.

Reply:
good call, fixed

Reply:
I notice you did away with this line `const logItem = Object.assign({}, messageLog[i])`. Was your intention to operate on a reference to the object or on a copy of the object?

Reply:
@mxsnyder This is probably no the right way to do this, but for times sake i did operate with the reference for some reason because it made it easier. I will log a bug to refactor this.

Reply:
#3